{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building communities\n",
    "\n",
    "`micom` will construct communities from a specification via a Pandas DataFrame. Here, the DataFrame needs at least two columns: \"id\" and \"file\" which specify the ID of the organism/tissue and a file containing the actual individual model. \n",
    "\n",
    "To make more sense of that we can look at a small example. `micom` comes with a function that generates a simple example community specification consisting of several copies of a small *E. coli* model containing only the central carbon metabolism."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>reactions</th>\n",
       "      <th>metabolites</th>\n",
       "      <th>file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia_coli_1</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Escherichia_coli_2</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Escherichia_coli_3</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Escherichia_coli_4</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Escherichia_coli_5</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id        genus          species  reactions  metabolites  \\\n",
       "0  Escherichia_coli_1  Escherichia  Eschericia coli         95           72   \n",
       "1  Escherichia_coli_2  Escherichia  Eschericia coli         95           72   \n",
       "2  Escherichia_coli_3  Escherichia  Eschericia coli         95           72   \n",
       "3  Escherichia_coli_4  Escherichia  Eschericia coli         95           72   \n",
       "4  Escherichia_coli_5  Escherichia  Eschericia coli         95           72   \n",
       "\n",
       "                                                file  \n",
       "0  /home/cdiener/code/micom/micom/data/e_coli_cor...  \n",
       "1  /home/cdiener/code/micom/micom/data/e_coli_cor...  \n",
       "2  /home/cdiener/code/micom/micom/data/e_coli_cor...  \n",
       "3  /home/cdiener/code/micom/micom/data/e_coli_cor...  \n",
       "4  /home/cdiener/code/micom/micom/data/e_coli_cor...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from micom.data import test_taxonomy\n",
    "\n",
    "taxonomy = test_taxonomy()\n",
    "taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we see this specification contains the required fields and some more information. In fact the specification may contain any number of additional information which will be saved along with the community model. One special example is \"abundance\" which we will get to know soon :)\n",
    "\n",
    "In order to convert the specification in a community model we will use the `Community` class from `micom` which derives from the cobrapy `Model` class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Build a community with a total of 495 reactions.\n"
     ]
    }
   ],
   "source": [
    "from micom import Community\n",
    "\n",
    "com = Community(taxonomy)\n",
    "print(\"Build a community with a total of {} reactions.\".format(len(com.reactions)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This includes the correctly scaled exchange reactions with the internal medium and initializes the external imports to the maximum found in all models. The original taxonomy is maintained in the `com.taxonomy` attribute.\n",
    "\n",
    "Note that `micom` can figure out how to read a variety of different file types based on the extension. It curently supports:\n",
    "\n",
    "- `.pickle` for pickled models\n",
    "- `.xml` or `.gz` for XML models\n",
    "- `.json` for JSON models\n",
    "- `.mat` for COBRAtoolbox models\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>reactions</th>\n",
       "      <th>metabolites</th>\n",
       "      <th>file</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Escherichia_coli_1</th>\n",
       "      <td>Escherichia_coli_1</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Escherichia_coli_2</th>\n",
       "      <td>Escherichia_coli_2</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Escherichia_coli_3</th>\n",
       "      <td>Escherichia_coli_3</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Escherichia_coli_4</th>\n",
       "      <td>Escherichia_coli_4</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Escherichia_coli_5</th>\n",
       "      <td>Escherichia_coli_5</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Eschericia coli</td>\n",
       "      <td>95</td>\n",
       "      <td>72</td>\n",
       "      <td>/home/cdiener/code/micom/micom/data/e_coli_cor...</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    id        genus          species  \\\n",
       "id                                                                     \n",
       "Escherichia_coli_1  Escherichia_coli_1  Escherichia  Eschericia coli   \n",
       "Escherichia_coli_2  Escherichia_coli_2  Escherichia  Eschericia coli   \n",
       "Escherichia_coli_3  Escherichia_coli_3  Escherichia  Eschericia coli   \n",
       "Escherichia_coli_4  Escherichia_coli_4  Escherichia  Eschericia coli   \n",
       "Escherichia_coli_5  Escherichia_coli_5  Escherichia  Eschericia coli   \n",
       "\n",
       "                    reactions  metabolites  \\\n",
       "id                                           \n",
       "Escherichia_coli_1         95           72   \n",
       "Escherichia_coli_2         95           72   \n",
       "Escherichia_coli_3         95           72   \n",
       "Escherichia_coli_4         95           72   \n",
       "Escherichia_coli_5         95           72   \n",
       "\n",
       "                                                                 file  \\\n",
       "id                                                                      \n",
       "Escherichia_coli_1  /home/cdiener/code/micom/micom/data/e_coli_cor...   \n",
       "Escherichia_coli_2  /home/cdiener/code/micom/micom/data/e_coli_cor...   \n",
       "Escherichia_coli_3  /home/cdiener/code/micom/micom/data/e_coli_cor...   \n",
       "Escherichia_coli_4  /home/cdiener/code/micom/micom/data/e_coli_cor...   \n",
       "Escherichia_coli_5  /home/cdiener/code/micom/micom/data/e_coli_cor...   \n",
       "\n",
       "                    abundance  \n",
       "id                             \n",
       "Escherichia_coli_1        0.2  \n",
       "Escherichia_coli_2        0.2  \n",
       "Escherichia_coli_3        0.2  \n",
       "Escherichia_coli_4        0.2  \n",
       "Escherichia_coli_5        0.2  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "com.taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can notice we have gained a new column called `abundance`. This column quantifies the relative quantity of each individual in the community. Since we did not specify this in the original taxonomy `micom` assumes that all individuals are present in the same quantity."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The presented community here is pretty simplistic. For microbial communities `micom` includes a larger taxonomy for 773 microbial species from the [AGORA paper](https://doi.org/10.1038/nbt.3703). Here the \"file\" column only contains the base names for the files but you can easily prepend any path as demonstrated in the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>oxygen_status</th>\n",
       "      <th>...</th>\n",
       "      <th>draft_created</th>\n",
       "      <th>platform</th>\n",
       "      <th>kbase_genome_id</th>\n",
       "      <th>pubseed_id</th>\n",
       "      <th>ncbi_id</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>genes</th>\n",
       "      <th>reactions</th>\n",
       "      <th>metabolites</th>\n",
       "      <th>file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abiotrophia defectiva ATCC 49176</td>\n",
       "      <td>Abiotrophia_defectiva_ATCC_49176</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Aerococcaceae</td>\n",
       "      <td>Abiotrophia</td>\n",
       "      <td>Abiotrophia defectiva</td>\n",
       "      <td>Facultative anaerobe</td>\n",
       "      <td>...</td>\n",
       "      <td>07/01/14</td>\n",
       "      <td>ModelSEED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Abiotrophia defectiva ATCC 49176 (592010.4)</td>\n",
       "      <td>592010.0</td>\n",
       "      <td>2041839</td>\n",
       "      <td>598</td>\n",
       "      <td>1069</td>\n",
       "      <td>840</td>\n",
       "      <td>models/Abiotrophia_defectiva_ATCC_49176.xml</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acidaminococcus fermentans DSM 20731</td>\n",
       "      <td>Acidaminococcus_fermentans_DSM_20731</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Acidaminococcales</td>\n",
       "      <td>Acidiaminococcaceae</td>\n",
       "      <td>Acidaminococcus</td>\n",
       "      <td>Acidaminococcus fermentans</td>\n",
       "      <td>Obligate anaerobe</td>\n",
       "      <td>...</td>\n",
       "      <td>04/17/16</td>\n",
       "      <td>Kbase</td>\n",
       "      <td>kb|g.2555</td>\n",
       "      <td>Acidaminococcus fermentans DSM 20731 (591001.3)</td>\n",
       "      <td>591001.0</td>\n",
       "      <td>2329769</td>\n",
       "      <td>646</td>\n",
       "      <td>1090</td>\n",
       "      <td>903</td>\n",
       "      <td>models/Acidaminococcus_fermentans_DSM_20731.xml</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acidaminococcus intestini RyC-MR95</td>\n",
       "      <td>Acidaminococcus_intestini_RyC_MR95</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Selenomonadales</td>\n",
       "      <td>Acidaminococcaceae</td>\n",
       "      <td>Acidaminococcus</td>\n",
       "      <td>Acidaminococcus intestini</td>\n",
       "      <td>Obligate anaerobe</td>\n",
       "      <td>...</td>\n",
       "      <td>08/03/14</td>\n",
       "      <td>ModelSEED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acidaminococcus intestini RyC-MR95 (568816.4)</td>\n",
       "      <td>568816.0</td>\n",
       "      <td>2487765</td>\n",
       "      <td>599</td>\n",
       "      <td>994</td>\n",
       "      <td>827</td>\n",
       "      <td>models/Acidaminococcus_intestini_RyC_MR95.xml</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acidaminococcus sp. D21</td>\n",
       "      <td>Acidaminococcus_sp_D21</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Selenomonadales</td>\n",
       "      <td>Acidaminococcaceae</td>\n",
       "      <td>Acidaminococcus</td>\n",
       "      <td>unclassified Acidaminococcus</td>\n",
       "      <td>Obligate anaerobe</td>\n",
       "      <td>...</td>\n",
       "      <td>06/29/12</td>\n",
       "      <td>ModelSEED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acidaminococcus sp. D21 (563191.3)</td>\n",
       "      <td>563191.0</td>\n",
       "      <td>2238973</td>\n",
       "      <td>598</td>\n",
       "      <td>851</td>\n",
       "      <td>768</td>\n",
       "      <td>models/Acidaminococcus_sp_D21.xml</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acinetobacter calcoaceticus PHEA-2</td>\n",
       "      <td>Acinetobacter_calcoaceticus_PHEA_2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Acinetobacter</td>\n",
       "      <td>Acinetobacter calcoaceticus</td>\n",
       "      <td>Aerobe</td>\n",
       "      <td>...</td>\n",
       "      <td>04/18/16</td>\n",
       "      <td>Kbase</td>\n",
       "      <td>kb|g.3519</td>\n",
       "      <td>Acinetobacter calcoaceticus PHEA-2 (871585.3)</td>\n",
       "      <td>871585.0</td>\n",
       "      <td>3862530</td>\n",
       "      <td>1026</td>\n",
       "      <td>1561</td>\n",
       "      <td>1165</td>\n",
       "      <td>models/Acinetobacter_calcoaceticus_PHEA_2.xml</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               organism                                    id  \\\n",
       "0      Abiotrophia defectiva ATCC 49176      Abiotrophia_defectiva_ATCC_49176   \n",
       "1  Acidaminococcus fermentans DSM 20731  Acidaminococcus_fermentans_DSM_20731   \n",
       "2    Acidaminococcus intestini RyC-MR95    Acidaminococcus_intestini_RyC_MR95   \n",
       "3               Acidaminococcus sp. D21                Acidaminococcus_sp_D21   \n",
       "4    Acinetobacter calcoaceticus PHEA-2    Acinetobacter_calcoaceticus_PHEA_2   \n",
       "\n",
       "    kingdom          phylum                class              order  \\\n",
       "0  Bacteria      Firmicutes              Bacilli    Lactobacillales   \n",
       "1  Bacteria      Firmicutes        Negativicutes  Acidaminococcales   \n",
       "2  Bacteria      Firmicutes        Negativicutes    Selenomonadales   \n",
       "3  Bacteria      Firmicutes        Negativicutes    Selenomonadales   \n",
       "4  Bacteria  Proteobacteria  Gammaproteobacteria    Pseudomonadales   \n",
       "\n",
       "                family            genus                       species  \\\n",
       "0        Aerococcaceae      Abiotrophia         Abiotrophia defectiva   \n",
       "1  Acidiaminococcaceae  Acidaminococcus    Acidaminococcus fermentans   \n",
       "2   Acidaminococcaceae  Acidaminococcus     Acidaminococcus intestini   \n",
       "3   Acidaminococcaceae  Acidaminococcus  unclassified Acidaminococcus   \n",
       "4        Moraxellaceae    Acinetobacter   Acinetobacter calcoaceticus   \n",
       "\n",
       "          oxygen_status                       ...                         \\\n",
       "0  Facultative anaerobe                       ...                          \n",
       "1     Obligate anaerobe                       ...                          \n",
       "2     Obligate anaerobe                       ...                          \n",
       "3     Obligate anaerobe                       ...                          \n",
       "4                Aerobe                       ...                          \n",
       "\n",
       "  draft_created   platform kbase_genome_id  \\\n",
       "0      07/01/14  ModelSEED             NaN   \n",
       "1      04/17/16      Kbase       kb|g.2555   \n",
       "2      08/03/14  ModelSEED             NaN   \n",
       "3      06/29/12  ModelSEED             NaN   \n",
       "4      04/18/16      Kbase       kb|g.3519   \n",
       "\n",
       "                                        pubseed_id   ncbi_id genome_size  \\\n",
       "0      Abiotrophia defectiva ATCC 49176 (592010.4)  592010.0     2041839   \n",
       "1  Acidaminococcus fermentans DSM 20731 (591001.3)  591001.0     2329769   \n",
       "2    Acidaminococcus intestini RyC-MR95 (568816.4)  568816.0     2487765   \n",
       "3               Acidaminococcus sp. D21 (563191.3)  563191.0     2238973   \n",
       "4    Acinetobacter calcoaceticus PHEA-2 (871585.3)  871585.0     3862530   \n",
       "\n",
       "  genes reactions  metabolites  \\\n",
       "0   598      1069          840   \n",
       "1   646      1090          903   \n",
       "2   599       994          827   \n",
       "3   598       851          768   \n",
       "4  1026      1561         1165   \n",
       "\n",
       "                                              file  \n",
       "0      models/Abiotrophia_defectiva_ATCC_49176.xml  \n",
       "1  models/Acidaminococcus_fermentans_DSM_20731.xml  \n",
       "2    models/Acidaminococcus_intestini_RyC_MR95.xml  \n",
       "3                models/Acidaminococcus_sp_D21.xml  \n",
       "4    models/Acinetobacter_calcoaceticus_PHEA_2.xml  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from micom.data import agora\n",
    "\n",
    "tax = agora.copy()\n",
    "tax.file = \"models/\" + tax.file  # assuming you have downloaded the AGORA models to the \"models\" folder\n",
    "tax.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
